lib/_http_agent.js

```diff
@@ -392,13 +418,27 @@ Agent.prototype.removeSocket = function removeSocket(s, options) {
         // Don't leak
         if (sockets[name].length === 0)
           delete sockets[name];
+        this.totalSocketCount--;
       }
     }
   }
 
+  let req;
   if (this.requests[name] && this.requests[name].length) {
     debug('removeSocket, have a request, make a socket');
-    const req = this.requests[name][0];
+    req = this.requests[name][0];
+  } else {
+    for (const prop in this.requests) {
+      debug('removeSocket, have a request with different origin,' +
+        ' make a socket');
+      req = this.requests[prop][0];
+      options = req._options;
+      break;
+    }
+  }
```

Comment:
Just a note here... this logic will not be FIFO across origins. There might be older requests in a different origin, but if the origin which releases the socket has pending requests that will be prioritized.

Not sure if that's a problem... just noting for reference... does this need to be documented?

Reply:
This is indeed a place to pay attention to, I will first documented.